opm/autodiff/CPRPreconditioner.hpp

```diff
@@ -43,8 +46,222 @@
 
 namespace Opm
 {
```

Comment:
Basically setting up the parallel solver is quite different from the sequential ones. They usually require sequential ones  with a wrapper that takes care about the necessary synchronization. 

The code is there to select and construct the correct types depending on whether we are running in parallel. Special care has to be taken to correctly decronstuct the wrapped objects as the wrapper only hold  a reference to them. 

The additional boiler plate code is needed to provide neat functions for constructing the preconditioners, etc and not clutter the main code.